### PR TITLE
Impro port of processImages transform

### PR DESCRIPTION
**Pull request by alexjeffburke**
This PR results in a working processImage() transform that operates via `impro`. There are a couple of tests that don't work, but I thought I might need a hand tracking down the issues.

### CODE REVIEW

**Comment by alexjeffburke** [comment]
@papandreou this port is inching closer to being usable - I've tracked down and fixed at least one issue with adding streams into the pipeline (used for the "custom filters"). If you have any spare bandwidth, would you mind eyeballing the rest of the failures? If I know what the issues are I'd be happy to patch `impro` and `express-processimage` further if need be, but I'm not well versed enough in what builder wants to achieve to do much more :)

**Comment by papandreou** [comment]
Great work, thanks a lot for doing this!

Looks like all the failing tests are related to `pngcrush`. I'll try to find out what's going on.

**Comment by papandreou** [comment]
In two of the tests pngcrush produces the following on stderr and doesn't write an output file :thinking:

```
   | pngcrush-1.8.0
   |    Copyright (C) 1998-2002, 2006-2016 Glenn Randers-Pehrson
   |    Portions copyright (C) 2005       Greg Roelofs
   | This is a free, open-source program.  Permission is irrevocably
   | granted to everyone to use this version of pngcrush without
   | payment of any fee.
   | Executable name is pngcrush
   | It was built with   bundled libpng-1.6.21
   | and is running with bundled libpng-1.6.21
   |    Copyright (C) 1998-2004, 2006-2016 Glenn Randers-Pehrson,
   |    Copyright (C) 1996, 1997 Andreas Dilger,
   |    Copyright (C) 1995, Guy Eric Schalnat, Group 42 Inc.,
   | and bundled zlib-1.2.8, Copyright (C) 1995-2013,
   |    Jean-loup Gailly and Mark Adler.
   | It was compiled with gcc version 4.8.4.
  
     Recompressing IDAT chunks in /tmp/12013-22704-1s6gs4u.ojowj.png
     Total length of data found in critical chunks           =      4945
  While converting /tmp/12013-22704-1s6gs4u.ojowj.png to /tmp/12013-22704-1tdiom6.jwwe.png:
    pngcrush caught libpng error:
     Wrote palette index exceeding num_palette
  
     Critical chunk length, method 151 (ws 14 fm 0 zl 4 zs 0) =      4872
     CPU time decoding 0.001, encoding 0.001, other 0.000, total 0.002 sec.
```

**Comment by papandreou** [comment]
Looked into the `processImages filter ordering should handle filters in order: optipng, pngquant, pngcrush` test case, where it looks like `pngcrush` choking on the output of `pngquant`. It's reproducible on the command line, so it's probably not related to impro after all:

```
$ <testdata/transforms/processImages/pngs/purplealpha24bit.png node_modules/.bin/pngquant - > quanted.png && pngcrush quanted.png quanted_and_crushed.png

 | pngcrush-1.8.0
 |    Copyright (C) 1998-2002, 2006-2016 Glenn Randers-Pehrson
 |    Portions copyright (C) 2005       Greg Roelofs
 | This is a free, open-source program.  Permission is irrevocably
 | granted to everyone to use this version of pngcrush without
 | payment of any fee.
 | Executable name is pngcrush
 | It was built with   bundled libpng-1.6.21
 | and is running with bundled libpng-1.6.21
 |    Copyright (C) 1998-2004, 2006-2016 Glenn Randers-Pehrson,
 |    Copyright (C) 1996, 1997 Andreas Dilger,
 |    Copyright (C) 1995, Guy Eric Schalnat, Group 42 Inc.,
 | and bundled zlib-1.2.8, Copyright (C) 1995-2013,
 |    Jean-loup Gailly and Mark Adler.
 | It was compiled with gcc version 4.8.4.

   Reading IHDR chunk.
   Reading gAMA chunk.
   Reading sRGB chunk.
   Reading tEXt chunk.
   Reading pHYs chunk.
   Reading PLTE chunk.
   Reading tRNS chunk.
   Reading IDAT chunk.
   Reading IEND chunk.
   Recompressing IDAT chunks in quanted.png
   Total length of data found in critical chunks           =      4945
While converting quanted.png to quanted_and_crushed.png:
  pngcrush caught libpng error:
   Wrote palette index exceeding num_palette

   Critical chunk length, method 151 (ws 14 fm 0 zl 4 zs 0) =      4872
   CPU time decoding 0.001, encoding 0.001, other 0.000, total 0.002 sec.
```

**Comment by alexjeffburke** [comment]
@papandreou interesting - do we think those tests are indicative of regressions in the tools? I want to peek at the double callback messages regardless but otherwise perhaps these shouldn’t block progress?

**Comment by papandreou** [comment]
The same error is reproducible with the cli tools on master, so maybe this is actually fixing a bug, but exposing another in pngcrush in the process -- will try to find out.

Maybe we should just kick pngcrush out of the "auto lossless" thing.

**Comment by papandreou** [comment]
I looked into it a bit now, and it looks like the problem is that on `impro-port`, `pngcrush` ends up being called without the `-noreduce` switch, which causes it to barf on the `IDAT` chunk as output by the other tools. Apparently we've previously discovered that `pngcrush` is pretty brittle, which has led to 1ac9a1209c93d19d5060ead9d45c13b079b58f26 and 70311aa82c8614e4da01e2cc5061b71d7f46045b

Adding `noreduce` to [the list of operations supported by the `pngcrush` engine](https://github.com/papandreou/impro/blob/ab9572884980e03b68c9fcd3e285c1e5362e2456/src/engines/pngcrush.js#L9) fixes all the breaking tests :relieved:

**Comment by alexjeffburke** [comment]
@papandreou https://github.com/papandreou/impro/commit/16f863875cf6697a75ab1bdc861cbc3bdf84ef1f and am pushing to the port with the bump now :)

**Comment by alexjeffburke** [comment]
Ok, we're down to a single failure! Seems the one of the expectations about an SVG -> PNG conversion has changed - do you know if that is problematic or is it just due to potententially being processed by different libraries given the engine selection etc?

**Comment by papandreou** [comment]
That one seems to be because `pngcrush` isn't properly instructed to remove the `pHYs` chunk, which [the test case clearly intends](https://github.com/assetgraph/assetgraph-builder/blob/1c12cbb6757f922b96abd7e790b035689c151c29/testdata/transforms/buildProduction/JavaScriptStaticUrlWithProcessedImage/index.html#L5).

Looks like `prepareImproQueryString` turns `resize=100,100&pngcrush=-rem+pHYs` into
`resize=100,100&pngcrush&rem+pHYs`, which makes it end up in `leftover`:

```js
{ operations:
   [ { name: 'resize', args: [Array] },
     { name: 'pngcrush', args: [] } ],
  leftover: 'rem+pHYs',
  consumed: 'resize=100,100&pngcrush' }
```

Is that one of the cases where the syntax is changing, or is there a bug in impro? :thinking:

**Comment by alexjeffburke** [comment]
That is super useful to help track it down, thanks - I don't think it's an improvement bug per se, but it's going to be an issue somewhere in the query string processing - either `impro.parse()`, but I have a sneaky feeling that more likely it is the "bridge" within `express-processimage`. Taking a look.. 👀

**Comment by alexjeffburke** [comment]
Success! 🎉 No user-facing syntax changes are required either, so I think it's time to release the dependent packages and then the only reason you need to bump the major is to drop node < 8 :)

**Comment by papandreou** [comment]
Great, thanks a lot for doing this! 🤗

**Comment by alexjeffburke** [comment]
I spoke a fraction too soon - went over the entire PR and there were a couple of thing to reenable after the porting work. The first test is around an output image not being found - I think that's probably because it's named a little different, but I don't know the `ag` APIs well enough to correct that. The second is the optipng argument not being applied - reason for that is we're expecting an argument in order for it to do anything.. so it is a little unclear to me what is desired by that test?

**Comment by papandreou** [comment]
The first one is just a file naming issue, I suggest fixing it like this:

```diff
diff --git a/test/transforms/processImages.js b/test/transforms/processImages.js
index 8a7fbcd..39e5657 100644
--- a/test/transforms/processImages.js
+++ b/test/transforms/processImages.js
@@ -373,7 +373,7 @@ describe('processImages', function() {
       )
       .queue(function(assetGraph) {
         var redAlpha24BitPngquanted = assetGraph.findAssets({
-          url: /\/redalpha24bit\.pngquant256\.png$/
+          fileName: 'redalpha24bit.pngquant.png'
         })[0];
         expect(
           _.toArray(redAlpha24BitPngquanted.rawSrc.slice(0, 4)),
```

The problem in the other test is that a query string of `optipng` gets parsed as:

```
{ operations: [], leftover: 'optipng', consumed: '' }
```

Seems like it needs to be fixed in e-p with something like:

```diff
diff --git a/lib/prepareImproQueryString.js b/lib/prepareImproQueryString.js
index 5add721..b7d1389 100644
--- a/lib/prepareImproQueryString.js
+++ b/lib/prepareImproQueryString.js
@@ -15,13 +15,12 @@ const resizeOptions = {
 };
 
 const queryStringEngineAndArgsRegex = new RegExp(
-  `^(${Object.keys(isSupportedEngine).join('|')})=(.*)`
+  `^(${Object.keys(isSupportedEngine).join('|')})(?:=(.*))?`
 );
 
 module.exports = function prepareQueryString(queryString) {
   const keyValuePairs = queryString.split('&');
   const queryStringFragments = [];
-
   let hasResize = false;
   let optionToResize;
 
@@ -29,7 +28,7 @@ module.exports = function prepareQueryString(queryString) {
     let m;
 
     if ((m = pair.match(queryStringEngineAndArgsRegex)) !== null) {
-      const [, engineName, engineArgs] = m;
+      const [, engineName, engineArgs = ''] = m;
       const result = [engineName];
       const splitChar = engineArgs.includes('+') ? '+' : ',';
       const remaining = engineArgs.split(splitChar);
```

**Comment by alexjeffburke** [comment]
Hmmm - I can’t remember what just specifying an engine in the query string like that will do on the `impro` side.. IIRC it just selects an engine - without an output format or an argument (seems optipng takes a level setting) it might just be silent passthrough. Do you know what this did before?

**Comment by papandreou** [comment]
For those optimization tools that don't _require_ any options to be specified (eg. `optipng` where `-o` is optional), it runs the image through that tool with whatever its default options are.

**Comment by alexjeffburke** [comment]
@papandreou over to you :)

File: package.json [hunk 1]
```diff
@@ -46,9 +46,10 @@
     "chalk": "^3.0.0",
     "esanimate": "^1.1.0",
     "estraverse": "^4.2.0",
-    "express-processimage": "^8.1.0",
+    "express-processimage": "^9.0.2",
```
[the remark this answers is not in the document]
Comment (alexjeffburke) [reply package.json]: Yeah, I think an `impro-querystring` module is likely way to go. My feeling is that both "impro native" and the legacy query string processing code should move there together.

There is a little extra `impro` work required though - first, if`createPipeline()` is called with a string that is automagically parsed and the output operations added at the moment. I think we should just hard split that and call it a day. Second, the native query string parser currently makes use of `validOperation()` and any `allowOperation` supplied by the user - we need to decide how to handle that when it is a separate package. I think I had the beginnings of that somewhere.

With the above as the moving plan, would you be ok with landing the PR as-is? It at least gets things working again on modern systems and I reckon it is a much better base to be able to follow-up on :)

[the remark this answers is not in the document]
Comment (papandreou) [reply package.json]: Sure! I'll merge.

**Pull request merged**